### PR TITLE
Add New OP: gumbel_softmax

### PR DESCRIPTION
**Pull request by YuanRisheng**
### PR types

New features
### PR changes

OPs
### Describe

Add new op named Gumbel Softmax for paddle.
This op is used for sampling from the Gumbel-Softmax distribution and optionally discretizes.

Verify the performance of Gumbel softmax operator locally. The input is random tensor (shape = [200,30]) and repeat the test 100 times. Compare  Paddle with Pytorch: the average time-consuming of Paddle is 0.0114 seconds, and the average time-consuming of Pytorch is 0.0123 seconds. The performance of Gumbel softmax operator of Paddle is about 7% better than that of Pytorch.

API:
        `paddle.nn.functional.gumbel_softmax()`

Usage example:

    import paddle
    import paddle.nn.functional as F

    logits = paddle.randn([4, 6])
    temperature = 0.01
    gumbel_softmax = F.gumbel_softmax(logits, temperature)
    print(gumbel_softmax)
    The result is as follows:
     [[0.00000001, 1.        , 0.00000000, 0.00000000, 0.00000006, 0.00000000],
     [0.00000000, 0.00000000, 0.00000000, 0.00000000, 0.00000000, 1.        ],
     [0.00000062, 0.00000000, 0.00000000, 0.00000000, 0.00000000, 0.99999940],
     [0.00000000, 0.00000000, 0.00000000, 0.00001258, 0.99998736, 0.00000000]]

Doc
Cn Doc PR: [PaddlePaddle/docs/3857](https://github.com/PaddlePaddle/docs/pull/3857)

image
image

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        paddle.enable_static()
+
+        def test_dtype():
+            with paddle.static.program_guard(paddle.static.Program()):
+                x_int32 = paddle.fluid.data(
+                    name='x_int32', shape=[2, 3], dtype='int32')
+                paddle.nn.functional.gumbel_softmax(x_int32)
+
+        self.assertRaises(TypeError, test_dtype)
```
Comment (MingMingShangTian) [inline python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py:225]: 这里单测缺少了
```python
if __name__ == '__main__':
    unittest.main()
```

Comment (YuanRisheng) [reply python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py:225]: done

File: paddle/fluid/operators/gumbel_softmax_op.cc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+class GumbelSoftmaxOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    return UnaryOpUnchangedInferShapeCheckAxis(ctx);
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.cc:31]: GetExpectedKernelType这里可以不写，只有一个输入X，默认根据X的特征确定Kernel类型

File: paddle/fluid/operators/gumbel_softmax_op.cc [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+             "each class.");
+    AddOutput("Out", "The sampled tensor with the same shape as X.");
+    AddAttr<float>("temperature",
+                   "(float, default 1.0) non-negative scalar temperature.")
+        .SetDefault(1.0);
+    AddAttr<bool>(
+        "hard",
+        "(bool, default false) "
+        "if True, the returned samples will be discretized as one-hot vectors, "
+        "but will be differentiated as if it is the soft sample in autograd")
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.cc]: autograd后面需要`.`?

File: paddle/fluid/operators/gumbel_softmax_op.cc [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+class GumbelSoftmaxGradOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("Out"), "Input", "Out", "gumbel_softmax_grad");
+    OP_INOUT_CHECK(ctx->HasInput(framework::GradVarName("Out")), "Input",
+                   "Out@grad", "gumbel_softmax_grad");
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.cc]: Out@grad -> Out@GRAD

File: paddle/fluid/operators/gumbel_softmax_op.cu [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+template <typename K, typename V>
+using KeyValuePair = cub::KeyValuePair<K, V>;
+
+template <typename T>
+struct UniformGenerator {
+  T min_, max_;
+  unsigned int seed_;
+  unsigned int offset_ = 0;
+  __host__ __device__ UniformGenerator(T min, T max, unsigned int seed)
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.cu]: `__host__ __device__` -> `HOSTDEVICE`, 有全局的宏定义

File: paddle/fluid/operators/gumbel_softmax_op.cu [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+#endif
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+#include <thrust/random.h>
+#include <thrust/transform.h>
+#include "paddle/fluid/framework/generator.h"
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/framework/operator.h"
+#include "paddle/fluid/memory/memcpy.h"
+#include "paddle/fluid/operators/gumbel_softmax_op.h"
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.cu]: 需要放到最上面，参见规范

https://zh-google-styleguide.readthedocs.io/en/latest/google-cpp-styleguide/headers/#include

File: paddle/fluid/operators/gumbel_softmax_op.cu [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    // add gumbel noise to X
+    const int thread_size = 512;
+    int64_t block_size = (size + thread_size) / thread_size;
+    AddGumbelNoiseCUDAKernel<
+        T><<<block_size, thread_size, 0, context.stream()>>>(
+        input_data, output_data, random_data, temperature, size);
+  }
+};
+
+#endif
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.cu:161]: 这个endif好像没有配对的ifdef

File: paddle/fluid/operators/gumbel_softmax_op.h [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    Out->mutable_data<T>(context.GetPlace());
+    if (Out->numel() == 0) {
+      return;
+    }
+
+    const int size_to_axis = SizeToAxis(axis, X->dims());
+    const int size_from_axis = SizeFromAxis(axis, X->dims());
+    Tensor X_noise_2d, Out_2d;
+    X_noise_2d.Resize({size_to_axis, size_from_axis});
+    Out_2d.ShareDataWith(*Out).Resize({size_to_axis, size_from_axis});
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.h:196]: op内部推荐使用ShareDataWith，这里是必须使用的吗

File: paddle/fluid/operators/gumbel_softmax_op.h [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* Out = context.Input<Tensor>("Out");
+    auto* dOut = context.Input<Tensor>(framework::GradVarName("Out"));
+    auto* dX = context.Output<Tensor>(framework::GradVarName("X"));
+    const int rank = dX->dims().size();
+    const int axis = CanonicalAxis(context.Attr<int>("axis"), rank);
+    int axis_dim = dX->dims()[axis];
+    // allocate memory on device.
+    dX->mutable_data<T>(context.GetPlace());
+    if (dX->numel() == 0) {
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.h:233]: 这里直接return是什么场景

File: paddle/fluid/operators/gumbel_softmax_op.h [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+      return;
+    }
+
+    const int size_to_axis = SizeToAxis(axis, dX->dims());
+    const int size_from_axis = SizeFromAxis(axis, dX->dims());
+    Tensor dX_2d, Out_2d, dOut_2d;
+    dX_2d.ShareDataWith(*dX).Resize({size_to_axis, size_from_axis});
```
Comment (chenwhql) [inline paddle/fluid/operators/gumbel_softmax_op.h:240]: 同上，为什么必须使用sharedatawith

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        x = np.random.uniform(0.1, 1, self.shape).astype(self.dtype)
+        out = np.zeros(self.shape).astype(self.dtype)
+        self.inputs = {'X': OpTest.np_dtype_to_fluid_dtype(x)}
+        self.outputs = {'Out': out}
+
+    def test_check_output(self):
+        self.check_output_customized(self.verify_output)
+
+    def test_check_grad(self):
+        self.check_grad(["X"], "Out", max_relative_error=0.01)
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py]: 这里0.01的精度设置是什么根据？有特殊原因吗？默认是0.005

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        z = (self.counts - expected) / np.sqrt((expected * (1 - self.probs)))
+        # A (lazy) approximate 99% two-sided test:
+        # occurs with prob alpha~>=0.01 if unbiased
+        self.assertLess(np.max(np.abs(z)).item(), 2.58)
+
+    def test_check_grad(self):
+        self.check_grad(["X"], "Out", max_relative_error=0.01)
+
+
+class TestGumbelSoftmaxOpGrad(unittest.TestCase):
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py:135]: 为什么需要单独写这个单测，理论上前面应该测了反向？

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+            np.allclose(x_hard.grad.numpy(), x_soft.grad.numpy()), True)
+        paddle.enable_static()
+
+
+class TestGumbelSoftmaxAPI(unittest.TestCase):
+    def setUp(self):
+        self.x_shape = [2, 3, 4, 5]
+        self.x = np.random.uniform(-1., 1., self.x_shape).astype(np.float32)
+        self.count_expected = 24
+        self.place = paddle.CUDAPlace(0) \
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py:165]: with cuda的话应该是CPU和CUDA都测试？没有CUDA的话也应该有CPU测试

**Review by chenwhql — COMMENTED** [review]
在PR描述中加上英文文档预览截图，并且关联到中文文档PR，可参考 https://github.com/PaddlePaddle/Paddle/pull/33278

File: paddle/fluid/operators/gumbel_softmax_op.cc (hunk 2 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.cc:31]: Done.

File: paddle/fluid/operators/gumbel_softmax_op.cc (hunk 3 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.cc]: Done.

File: paddle/fluid/operators/gumbel_softmax_op.cc (hunk 4 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.cc]: Done.

File: paddle/fluid/operators/gumbel_softmax_op.cu (hunk 5 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.cu]: Done.

File: paddle/fluid/operators/gumbel_softmax_op.cu (hunk 6 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.cu]: Done.

File: paddle/fluid/operators/gumbel_softmax_op.h (hunk 9 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.h:233]: 这里的逻辑和softmax是一样的，所以也同样按照softmax算子的做法添加了对dx的边界处理

File: paddle/fluid/operators/gumbel_softmax_op.cu (hunk 7 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.cu:161]: 与`#if defined(__NVCC__) || defined(__HIPCC__)`配对

File: paddle/fluid/operators/gumbel_softmax_op.h (hunk 8 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.h:196]: 这里需要将out的数据resize到2维进行处理计算结果，使用shareDataWith可以减少copy次数

File: paddle/fluid/operators/gumbel_softmax_op.h (hunk 10 above)
Comment (YuanRisheng) [reply paddle/fluid/operators/gumbel_softmax_op.h:240]: 这里需要将out的数据resize到2维进行处理计算结果，使用shareDataWith可以减少copy次数

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py (hunk 11 above)
Comment (YuanRisheng) [reply python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py]: 由于梯度计算逻辑和softmax算子是一样的，这里精度误差0.01取自softmax算子使用的精度误差

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py (hunk 12 above)
Comment (YuanRisheng) [reply python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py:135]: 这里测的hard为ture和false条件下，反向梯度是否一致，和前边测的反向内容略有不同

File: python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py (hunk 13 above)
Comment (YuanRisheng) [reply python/paddle/fluid/tests/unittests/test_gumbel_softmax_op.py:165]: 这个下边代码是有的，没有cuda的话是cpu测试

**Review by chenwhql — DISMISSED** [review]
LGTM

**Review by lanxianghit — DISMISSED** [review]

**Review by zhhsplendid — DISMISSED** [review]
LGTM on ShareDataWith

**Review by chenwhql — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]
LGTM for API change

**Review by Xreki — APPROVED** [review]
LGTM for op benchmark ci

**Review by zhhsplendid — APPROVED** [review]
LGTM for ShareDataWith

**Review by TCChenlong — APPROVED** [review]
